### PR TITLE
[SPARK-37015][PYTHON] Inline type hints for python/pyspark/streaming/dstream.py

### PR DESCRIPTION
**Pull request by dchvn**
### What changes were proposed in this pull request?
Inline type hints for python/pyspark/streaming/dstream.py

### Why are the changes needed?
We can take advantage of static type checking within the functions by inlining the type hints.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing tests

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @itholic mind reviewing this please?

File: python/pyspark/streaming/dstream.py [hunk 1]
```diff
@@ -51,122 +76,165 @@ class DStream(object):
      - A function that is used to generate an RDD after each time interval
     """
 
-    def __init__(self, jdstream, ssc, jrdd_deserializer):
+    def __init__(
+        self,
+        jdstream: JavaObject,
+        ssc: "StreamingContext",
+        jrdd_deserializer: "Serializer",
```
Comment (itholic) [inline python/pyspark/streaming/dstream.py:83]: Same as https://github.com/apache/spark/pull/34293/files#r787298665.

Would you address the all such cases below??

Comment (zero323) [reply python/pyspark/streaming/dstream.py:83]: For starters, this should be re-synced with master, as covariance has some impact on annotations on implementation (I took over SPARK-37154 from Byron to deal with that on `pyspark.rdd`).

**Comment by zero323** [comment]
This should be resynced with master and reformatted.

**Comment by itholic** [comment]
Also mind taking a last look for this, @zero323 ?? 🙏

File: python/pyspark/streaming/dstream.py [hunk 2]
```diff
@@ -19,19 +19,46 @@
...
+    Optional,
+    Tuple,
+    TypeVar,
+    Union,
+    TYPE_CHECKING,
+    cast,
+    overload,
+)
+
+from py4j.java_gateway import JavaObject
```
Comment (zero323) [inline python/pyspark/streaming/dstream.py:38]: Normally we import these in `TYPE_CHECKING` blocks.

File: python/pyspark/streaming/dstream.py [hunk 3]
```diff
@@ -19,19 +19,46 @@
...
 from py4j.protocol import Py4JJavaError
 
-from pyspark import RDD
 from pyspark.storagelevel import StorageLevel
 from pyspark.streaming.util import rddToFileName, TransformFunction
 from pyspark.rdd import portable_hash
+from pyspark import RDD
```
Comment (zero323) [inline python/pyspark/streaming/dstream.py:44]: Let's avoid unnecessary moves.

File: python/pyspark/streaming/dstream.py [hunk 4]
```diff
@@ -51,125 +78,170 @@ class DStream:
...
-    def count(self):
+    def count(self) -> "DStream[int]":
         """
         Return a new DStream in which each RDD has a single element
         generated by counting each RDD of this DStream.
         """
         return self.mapPartitions(lambda i: [sum(1 for _ in i)]).reduce(operator.add)
 
-    def filter(self, f):
+    def filter(self, f: Callable[[T_co], bool]) -> "DStream[T_co]":
```
Comment (zero323) [inline python/pyspark/streaming/dstream.py:107]: In general, we should avoid using covariant types in arguments, unless necessary. See how it is done in `pyspark.rdd.RDD`.

File: python/pyspark/streaming/dstream.py [hunk 5]
```diff
@@ -51,125 +78,170 @@ class DStream:
...
+    def reduce(self, func: Callable[[T_co, T_co], T_co]) -> "DStream[T_co]":
         """
         Return a new DStream in which each RDD has a single element
         generated by reducing each RDD of this DStream.
         """
-        return self.map(lambda x: (None, x)).reduceByKey(func, 1).map(lambda x: x[1])
+        return (
+            self.map(lambda x: (None, x))  # type: ignore[misc]
+            .reduceByKey(func, 1)  # type: ignore[arg-type]
+            .map(lambda x: cast(T_co, x[1]))
```
Comment (zero323) [inline python/pyspark/streaming/dstream.py:176]: As above, you shouldn't need ignores here if you switch to non-covariant typevars.

File: python/pyspark/streaming/dstream.py [hunk 6]
```diff
@@ -192,40 +264,42 @@ def takeAndPrint(time, rdd):
...
 
-    def mapValues(self, f):
+    def mapValues(self: "DStream[Tuple[K, V]]", f: Callable[[V], U]) -> "DStream[Tuple[K, U]]":
         """
         Return a new DStream by applying a map function to the value of
         each key-value pairs in this DStream without changing the key.
         """
 
-        def map_values_fn(kv):
+        def map_values_fn(kv: Any) -> Any:
```
Comment (zero323) [inline python/pyspark/streaming/dstream.py:273]: This should be properly annotated.

File: python/pyspark/streaming/dstream.py [hunk 7]
```diff
@@ -254,31 +328,35 @@ def checkpoint(self, interval):
...
         return self.transform(lambda rdd: rdd.groupByKey(numPartitions))
 
-    def countByValue(self):
+    def countByValue(self) -> "DStream[Tuple[T_co, int]]":
         """
         Return a new DStream in which each RDD contains the counts of each
         distinct value in each RDD of this DStream.
         """
-        return self.map(lambda x: (x, 1)).reduceByKey(lambda x, y: x + y)
+        return self.map(lambda x: (x, 1)).reduceByKey(  # type: ignore[return-value, misc]
```
Comment (zero323) [inline python/pyspark/streaming/dstream.py:349]: Ditto.

File: python/pyspark/streaming/dstream.py [hunk 8]
```diff
@@ -316,15 +405,41 @@ def transform(self, func):
...
+    ) -> "DStream[V]":
+        ...
+
+    def transformWith(
+        self,
+        func: Union[
+            Callable[[RDD[T_co], RDD[U]], RDD[V]],
+            Callable[[datetime, RDD[T_co], RDD[U]], RDD[V]],
+        ],
+        other: RDD[U],
```
Comment (zero323) [inline python/pyspark/streaming/dstream.py:440]: This should actually be `DStream[U]`.

File: python/pyspark/streaming/dstream.py [hunk 9]
```diff
@@ -316,15 +405,41 @@ def transform(self, func):
...
 
         assert func.__code__.co_argcount == 2, "func should take one or two arguments"
         return TransformedDStream(self, func)
 
-    def transformWith(self, func, other, keepSerializer=False):
+    @overload
+    def transformWith(
+        self,
+        func: Callable[[RDD[T_co], RDD[U]], RDD[V]],
+        other: RDD[U],
```
Comment (zero323) [inline python/pyspark/streaming/dstream.py:420]: This should actually be `DStream[U]`.

File: python/pyspark/streaming/dstream.py [hunk 10]
```diff
@@ -316,15 +405,41 @@ def transform(self, func):
...
+        keepSerializer: bool = ...,
+    ) -> "DStream[V]":
+        ...
+
+    @overload
+    def transformWith(
+        self,
+        func: Callable[[datetime, RDD[T_co], RDD[U]], RDD[V]],
+        other: RDD[U],
```
Comment (zero323) [inline python/pyspark/streaming/dstream.py:429]: Ditto

File: python/pyspark/streaming/dstream.py [hunk 11]
```diff
@@ -333,33 +448,38 @@ def transformWith(self, func, other, keepSerializer=False):
...
-        jfunc = TransformFunction(self._sc, func, self._jrdd_deserializer, other._jrdd_deserializer)
+        jfunc = TransformFunction(
+            self._sc,
+            func,
+            self._jrdd_deserializer,
+            other._jrdd_deserializer,
+        )
         dstream = self._sc._jvm.PythonTransformed2DStream(
-            self._jdstream.dstream(), other._jdstream.dstream(), jfunc
+            self._jdstream.dstream(), other._jdstream.dstream(), jfunc  # type: ignore[attr-defined]
```
Comment (zero323) [inline python/pyspark/streaming/dstream.py:464]: Shouldn't be necessary when  signature is fixed. Also, similar cases in the following methods.

**Comment by zero323** [comment]
> Also mind taking a last look for this, @zero323 ?? 🙏

Sorry for the delay @itholic. Done.

**Comment by itholic** [comment]
No worries!! Thanks for the review, @zero323

**Comment by zero323** [comment]
@dchvn Please remember to re-sync this with master  (with SPARK-37014 merged, some of the ignores will be obsolete).

**Comment by itholic** [comment]
Looks good except the remaining comments.

Thanks for your time to contribution!

**Comment by dchvn** [comment]
Thanks, @zero323 , @itholic for your reviews!

**Comment by dchvn** [comment]
check the tests at my repo: https://github.com/dchvn/spark/actions/runs/2166305834

**Comment by zero323** [comment]
> check the tests at my repo: https://github.com/dchvn/spark/actions/runs/2166305834

There are still quite a few usages of covariant types that have to be addressed.

**Comment by zero323** [comment]
> check the tests at my repo: https://github.com/dchvn/spark/actions/runs/2166305834

There are still quite a few usages of covariant types that have to be addressed.

**Comment by dchvn** [comment]
> > check the tests at my repo: https://github.com/dchvn/spark/actions/runs/2166305834
> 
> There are still quite a few usages of covariant types that have to be addressed.

I'll take a look at the overview again, thanks

**Comment by zero323** [comment]
At the moment, there are at least few things left that should be fixed, with `updateStateByKey` being a blocker, as it fails in type check.

```python
diff --git a/python/pyspark/streaming/dstream.py b/python/pyspark/streaming/dstream.py
index 5aff442aa6..934b3ae578 100644
--- a/python/pyspark/streaming/dstream.py
+++ b/python/pyspark/streaming/dstream.py
@@ -479,7 +479,7 @@ class DStream(Generic[T_co]):
         """
         return self._jdstream.dstream().slideDuration().milliseconds() / 1000.0
 
-    def union(self, other: "DStream[U]") -> "DStream[Union[T, U]]":
+    def union(self: "DStream[T]", other: "DStream[U]") -> "DStream[Union[T, U]]":
         """
         Return a new DStream by unifying data of another DStream with this DStream.
 
@@ -491,15 +491,13 @@ class DStream(Generic[T_co]):
         """
         if self._slideDuration != other._slideDuration:
             raise ValueError("the two DStream should have same slide duration")
-        return self.transformWith(
-            lambda a, b: a.union(b), other, True  # type: ignore[arg-type, return-value]
-        )
+        return self.transformWith(lambda a, b: a.union(b), other, True)
 
     def cogroup(
         self: "DStream[Tuple[K, V]]",
         other: "DStream[Tuple[K, U]]",
         numPartitions: Optional[int] = None,
-    ) -> "DStream[Tuple[K, Tuple[List[V], List[U]]]]":
+    ) -> "DStream[Tuple[K, Tuple[ResultIterable[V], ResultIterable[U]]]]":
         """
         Return a new DStream by applying 'cogroup' between RDDs of this
         DStream and `other` DStream.
@@ -509,9 +507,7 @@ class DStream(Generic[T_co]):
         if numPartitions is None:
             numPartitions = self._sc.defaultParallelism
         return self.transformWith(
-            lambda a, b: a.cogroup(  # type: ignore[arg-type]
-                b, numPartitions
-            ),  # type: ignore[return-value]
+            lambda a, b: a.cogroup(b, numPartitions),
             other,
         )
 
@@ -673,7 +669,7 @@ class DStream(Generic[T_co]):
 
     def countByWindow(
         self: "DStream[T]", windowDuration: int, slideDuration: int
-    ) -> "DStream[Tuple[T, int]]":
+    ) -> "DStream[int]":
         """
         Return a new DStream in which each RDD has a single element generated
         by counting the number of elements in a window over this DStream.
@@ -682,7 +678,7 @@ class DStream(Generic[T_co]):
         This is equivalent to window(windowDuration, slideDuration).count(),
         but will be more efficient if window is large.
         """
-        return self.map(lambda x: 1).reduceByWindow(  # type: ignore[return-value]
+        return self.map(lambda x: 1).reduceByWindow(
             operator.add, operator.sub, windowDuration, slideDuration
         )
 
@@ -831,7 +827,7 @@ class DStream(Generic[T_co]):
         self: "DStream[Tuple[K, V]]",
         updateFunc: Callable[[Iterable[V], Optional[S]], S],
         numPartitions: Optional[int] = None,
-        initialRDD: Optional[RDD[Tuple[K, S]]] = None,
+        initialRDD: Optional[Union[RDD[Tuple[K, S]], Iterable[Tuple[K, S]]]] = None,
     ) -> "DStream[Tuple[K, S]]":
         """
         Return a new "state" DStream where the state for each key is updated by applying
@@ -865,7 +861,7 @@ class DStream(Generic[T_co]):
             self._jrdd_deserializer,
         )
         if initialRDD:
-            initialRDD = initialRDD._reserialize(self._jrdd_deserializer)
+            initialRDD = cast(RDD[Tuple[K, S]], initialRDD)._reserialize(self._jrdd_deserializer)
             assert self._sc._jvm is not None
             dstream = self._sc._jvm.PythonStateDStream(
                 self._jdstream.dstream(),

```

Other ignores, might require further look, but are not a blocker.

**Comment by dchvn** [comment]
updated, Sorry about the delay because of my weekend.

**Comment by itholic** [comment]
No worries! Thanks for fixing :-)

Seems like the test is passed at https://github.com/dchvn/spark/actions/runs/2182508677.

**Review by zero323 — APPROVED** [review]

**Comment by zero323** [comment]
Merged into master and branch-3.3

**Comment by dchvn** [comment]
Thanks, @zero323 @itholic

**Comment by itholic** [comment]
Thanks, @dchvn too

**Pull request closed without merging**